draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +391,343 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
+5. expected_tag = MAC(auth_key, concat(mode, envelope_nonce, inner_env, cleartext_creds))
+6. If !ct_equal(envelope.auth_tag, expected_tag),
+     raise MacError
+7. Output (client_private_key, export_key)
+~~~
+
+## Envelope Modes {#envelope-modes}
+
+The `EnvelopeMode` specifies the structure and encoding of the corresponding `InnerEnvelope`. This document specifies the values of the two aformentionned modes:
 
 ~~~
 enum {
-  base(1),
-  custom_identifier(2),
-  (255)
+ internal(1),
+ external(2),
+ (255)
 } EnvelopeMode;
 ~~~
 
-The `base` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+Each `EnvelopeMode` defines its own `InnerEnvelope` structure and must implement the following interface:
+- `inner_env, client_public_key = BuildInnerEnvelope(prk, nonce, client_private_key)`: build and return the mode's `InnerEnvelope` structure and the client's public key.
+- `client_private_key, client_public_key = RecoverKeys(prk, nonce, inner_env)`: recover and return the client's private and public keys for the AKE protocol.
+
+### Internal Key mode {#internal-mode}
+
+In this mode, the client's private and public keys are deterministically derived from the OPRF output. 
+
+With the internal key mode the `EnvelopeMode` value MUST be `internal` and the `InnerEnvelope` is empty, and the size `Ne` of the serialized `Envelope` is 1 + Nn + Nm.
+
+To generate the private key OPAQUE-3DH implements `DeriveAkeKeyPair(seed)` as follows:
 
 ~~~
-struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+DeriveAkeKeyPair(seed)
 
-struct {
-  opaque server_public_key[Npk];
-} CleartextCredentials;
+Parameter:
+- dst, domain separation tag for HashToScalar set to "OPAQUE-HashToScalar"
+
+Input:
+- seed, pseudo-random byte sequence used as a seed
+
+Output:
+- private_key, a private key
+- public_key, the associated public key
+
+Steps:
+1. private_key = HashToScalar(seed, dst)
+2. public_key = private_key * G
+3. Output (private_key, public_key)
 ~~~
 
-The `custom_identifier` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+HashToScalar(msg, dst): Hash arbitrary input to a scalar in the field of the AKE group as defined in {{I-D.irtf-cfrg-voprf}}, except that dst = "OPAQUE-HashToScalar".
 
 ~~~
-struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+BuildInnerEnvelope(prk, nonce, client_private_key)
 
-struct {
-  opaque server_public_key[Npk];
-  opaque client_identity<0..2^16-1>;
-  opaque server_identity<0..2^16-1>;
-} CleartextCredentials;
+Input:
+- prk, pseudo-random key
+- nonce, a unique nonce of length `Nn`
+- client_private_key, empty value. Not used in this function, it only serves to comply with the API
+
+Output:
+- inner_env, nil value (serves to comply with the API)
+- client_public_key, the client's AKE public key
+
+Steps:
+1. seed = Expand(prk, concat(nonce, "PrivateKey"), Nsk)
```

Comment:
Why is the length of the seed that is passed as input to DeriveAkeKeyPair of length `Nsk`?

Shouldn't it be `Nh`?

cc: @chris-wood

Reply:
Yes, you're right !

Reply:
ftr: reverting this change, as the amount of entropy in the seed should be no less than the amount of entropy in a private key